---
layout: lesson
root: .  # Is the only page that doesn't follow the pattern /:path/index.html
permalink: index.html  # Is the only page that doesn't follow the pattern /:path/index.html
---

Here you can find information about the advanced module "Viromics Bioinformatics" of the Masters program in Microbiology of the Friedrich-Schiller University of Jena.

It consists of 60 hours dry lab work to be completed in two weeks.

You will learn how to analyse bacteriophage sequences with computational approaches, covering the following topics:
  - Sequencing quality control
  - Assembly of long-read sequences
  - Identifying viral contigs
  - Open reading frame (ORF) predictions
  - Gene annotation
  - Host prediction
  - Viral taxonomy and phylogeny
  - Visualizations

At the end of the module, you will have the opportunity to present your work to your classmates and the teachers.

If you do not have a solid background in linux/bash, please follow this [self-study material](https://github.com/waltercostamb/course_viral-microbiology_2023/blob/main/study_material.md) before the start of the module.

At the top of the webpage, under "Extras", you will find notes on:  
- [writing a good lab book](https://mgxlab.github.io/Viromics2024/labbook/index.html)
- [accessing the HPC cluster Draco](https://mgxlab.github.io/Viromics2024/draco/index.html)
- [submitting jobs on draco](https://mgxlab.github.io/Viromics2024/sbatch/index.html)
- [setting up a python virtual environment](https://mgxlab.github.io/Viromics2024/virtualenv/index.html)
- [ideas for projects](https://mgxlab.github.io/Viromics2024/draco/index.html)

{% include base_path.html %}
<p align="center">
    <a href="{{ site.carpentries_site }}"><img src="{{ relative_root_path }}/assets/img/slide_module_2024.png" alt="Viromics workflow" width="800" /></a>
</p>

{% include links.md %}
