---
title: Getting to know your sequence data
teaching: 180
exercises: 0
questions: []
objectives:
- Understand the value of sequencing quality control
- Discuss how low quality reads affect genomic analysis
- Perform quality check on long-read data by creating plots
- Interpret quality check results
- Perform quality control on reads
keypoints:
---

## Assessing Sequencing Quality

> ## discussion
>
> Why do we need quality control?  
> 
> What is the impact of including low quality reads in downstream analyses?  
>
> What are some metrics for assessing sequencing quality?  
> 
> {: .source}
{: .discussion}

 [Nanopore sequencing quality] (https://help.nanoporetech.com/en/articles/6629615-where-can-i-find-out-more-about-quality-scores)

We will use [NanoPlot](https://github.com/wdecoster/NanoPlot?tab=readme-ov-file) to assess the quality of our reads. 

> ## Exercise
>
> Run NanoPlot on your reads
> - play around with the features
> - Use fastq files and summary.txt
> - Assess the sequencing quality for each sample
> 
>```bash
> NanoPlot -t ? --plots ? --color ? --fastq barcode1.fastq.gz -o output_dir/barcode1
>```
> 
> {: .source}
{: .challenge}


> ## sbatch script for NanoPlot
> ```bash
> #!/bin/bash
>#SBATCH --tasks=1
>#SBATCH --cpus-per-task=10
>#SBATCH --partition=short,standard
>#SBATCH --mem=1G
>#SBATCH --time=2:00:00
>#SBATCH --job-name=sequencing_QC
>#SBATCH --output=10_nanoplot/nanoplot.slurm.out.%j
>#SBATCH --error=10_nanoplot/nanoplot.slurm.err.%j
>
># First, we check the quality of the reads using nanoplot
># activate the conda environment containing nanoplot
>
>source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh && conda activate nanoplot_v1.41.3
>
>
># Run nanoplot on 3 fastq files in a for loop
># -t : threads (cpus)
># --plots : types of plots to produce, see gallery: https://gigabaseorgigabyte.wordpress.com/2017/06/01/example-gallery-of-nanoplot/ 
># --N50 : draw the N50 vertical line on the plots
># --color : color for the plots
># -o : output directory
># to see other options for colors and prefilters, run : NanoPlot -h
>
>for fn in ../data/sequences/*.fastq.gz
>do
>
>base_f=$(basename $fn .fastq.gz)        # select just the first part of the name
>NanoPlot -t 10 --plots dot --N50 --color slateblue --fastq $fn -o 10_nanoplot/${base_f}/
>
>done
>```
> {: .source}
{: .solution}


- Do we need to filter any reads?
- Filter reads using [Chopper](https://github.com/wdecoster/chopper) 


**Sequencing Quality Questions**
    - For illumina
        - phred scores
        - read quality disribution
    - for nanopore?
        - read quality 
        - read lengths
        - quality v. length
        - Do we need to remove quality reads?
