---
title: "Sequencing Quality I"
teaching: 120
exercises: 30
questions:
- "many for sure"
objectives:
- "does this work?"
keypoints:
---

# Download sequences and evaluate sequencing quality

Watch the lecture below and write down at least 3 questions and/or discussion point about it:

https://www.youtube.com/watch?v=OMzshfgJ4PQ  

- Click on the image to see lecture "Basic file formats in bioinformatics", by Prof Bas E. Dutilh:  
[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/D4WDdAbZW1Y/0.jpg)](https://www.youtube.com/watch?v=D4WDdAbZW1Y)

Copy viromics sequences following the teacher's directions. These will be used in your project.        

Answer the questions below using bash. Useful bash commands, by Varada Khot, can be found here: https://github.com/vmkhot/useful-scripts/blob/main/Linux%20Commands%20Cheat%20Sheet.md  

- how many lines does your FASTQ file has?
- how many sequences are there in your FASTQ file?
- Print the first 10 lines of your file in the terminal
- Print the last 10 lines of your file in the terminal
- Describe what you see in these lines and what they mean
- What is the GC content of your files?
- Create a plot of frequency (y-axis) versus GC content (x-axis) and describe the distribution. Use python (seaborn distplot, matplotlib hist) or R (hist or ggplot geom_hist)
- Evaluate the sequence quality following: https://github.com/vmkhot/Metagenome-workflows/blob/main/Nanopore-Long-Reads/Quality%20Control.md



<details> 
  <summary>
    > Q1: What is the best Language in the World? </summary>
  A1: JavaScript 
</details>

> ## callout
>
> text
> text
> text
>
> ~~~
> code
> ~~~
> {: .source}
{: .callout}

> ## challenge
>
> text
> text
> text
>
> ~~~
> code
> ~~~
> {: .source}
{: .challenge}


> ## Solution here
>
> text
> text
> text
>
> ~~~
> code
> ~~~
> {: .source}
{: .solution}


> ## discussion
>
> text
> text
> text
>
> ~~~
> code
> ~~~
> {: .source}
{: .discussion}

