---
title: Getting to know your files
teaching: 120
exercises: 0
questions: []
objectives:
- Interpret different file formats used in bioinformatics
- Utilize bash commands to look into sequence data
- Get a basic understanding of what sequence data looks like
- Work with sequence data in Python or R to create basic plots
keypoints:
- Sequence data and information about genes and genomes can come in many different formats
- The most common file formats are fasta (nucl. and amino acid), fastq, sam and bam, genbank, gff and tsv files
---

## Submitting jobs on Draco

When you login to Draco, you are on the "login node" and this is not the best place to run any programs or heavier scripts. You **MUST** be on a "compute node" to run scripts and programs

You can either request for resources from a compute node and run programs interactively or submit a job to the job scheduler, which then sends it to a compute node to complete.

[See here for what slurm architecture looks like](https://mgxlab.github.io/Viromics2024/draco/index.html)

You can either request resources from a node for an "interactive" shell or you can submit via sbatch

To get resources - [see here](https://mgxlab.github.io/Viromics2024/sbatch/index.html#salloc)

To submit a job, you have to make a script `my_slurm_script.sh`  and then submit it as `sbatch my_slurm_script.sh` . Detailed information on creating these scripts, including descriptions can be found in the "extras" [here](https://mgxlab.github.io/Viromics2024/sbatch/index.html)

[Example script here](https://mgxlab.github.io/Viromics2024/sbatch/index.html#example-sbatch-script)

## Navigating your filesystem space

> ## Exercise
> Take a look at the file structure and files. We have given you a subset of viromics data. Some helpful commands might be
> ```bash
> # list
> ls
> ll
>
> # path to working directory
> pwd
>
> # print first 10 lines
> head -10
> # print last 10 lines
> tail -10
>
> # open file on terminal (press 'q' to exit less)
> less
> more
>
> # count number of lines
> wc -l file.txt
>
> # search for "@" inside file 
> grep "@" file.fastq
>
> # count number of sequence headers in a fasta file
> grep -c ">" file.fasta
>
> ```
> {: .source}
{: .challenge}


[See more useful commands and one-liners here](https://github.com/vmkhot/useful-scripts/blob/main/Linux%20Commands%20Cheat%20Sheet.md#linux-commands-cheat-sheet)


## Understanding bioinformatics file formats

It is always a good idea to check the contents of your data files and make regular "sanity checks" to see if you understand everything. The first video covers different file formats commonly used in bioinformatics. Namely, FASTA, FASTQ, BAM, SAM, VCF, BCF, GFF, GTF and BED files (9 minutes).

[![Introduction to common file formats in bioinformatics & computational biology](https://img.youtube.com/vi/KZ2wqKFerG0/0.jpg)](https://www.youtube.com/watch?v=KZ2wqKFerG0&ab_channel=edu-ome)

The second video covers different types of sequence files, including fasta and fastq, phred scores, files containing metadata (tsv files) and compressed files (11 minutes).

[![Basic file formats in bioinformatics](https://img.youtube.com/vi/D4WDdAbZW1Y/0.jpg)]( https://www.youtube.com/watch?v=D4WDdAbZW1Y&ab_channel=BasE.Dutilh)

> ## Exercise
>
> 1. What information is contained in sequencing files?  
>     a) Choose 3 file formats and describe them
> 2. Find a fastq file in the folder with sequencing data on Draco.  
>     a) How many lines does this file have?  
>     b) How many sequences are present in the file?  
> 3. Print the first 10 lines and the last 10 lines in the terminal.  
>     a) What differences do you observe between these lines?  
> {: .source}
{: .challenge}
