---
start: True
title: Course Introduction
teaching: 60
exercises: 0
questions: []
objectives:
  - Understand course scope and evaluation criteria
  - Start and plan an effective lab book
  - login into Draco
keypoints:
- In this course you will analyze viral sequence data
- You need to keep a lab notebook
- You need to be able to access the HPC Draco
---

In this module you will learn how to analyse bacteriophage sequences with computational approaches. We will be using viromics sequence data from Dr. Xiu Jia (a Post Doc in our group). This is data was generated by sequencing samples from an estuary transect near Rostock. The samples were filtered through a 0.22um filter and then sequenced on a [Oxford Nanopore Technologies Flow Cell](https://nanoporetech.com/platform/technology) (long read sequencing). This filtration technique is commonly used to filter for viral particles. 

During the course, we will cover the following topics
- Sequencing quality control
- Assembly of long-read sequences
- Identifying viral contigs
- Open reading frame (ORF) predictions
- Gene annotation
- Host prediction
- Viral taxonomy and phylogeny
- Visualizations

**The theoretical parts** of the course are covered in the mornings. This includes reading relevant papers, watching video lectures, discussions on the concepts and tools. Document any questions and discussion in your lab book. This will also be considered towards your final grade.

**The practical parts** of the course will be in the afternoons, where you will get to use specialized bioinformatics tools for each of parts. These practical parts will also include data interpretation and comparisons. We will be available to help and guide you during this time.

**Visualization homework** will be assigned for some of the parts in lieu of the missed day. It will be up to you to complete this in your own time and to include it in your lab book and presentations

**The final presentation** will be held on September xx, 2024. This is where you can show off your understanding of the material and also what interesting things you identified in the data. 

**The final evaluation comprises:**

- Performance in the dry lab (participation, questions, asking for help and helping others) + lab journal (60%)
- Presentation of a final project (40%)

## Lab Notebook

Documenting your work is crucial in Computational Biology/Bioinformatics. This way you can make sure your work is reproducible, you can transfer text to other documents, such as reports or papers and you can send it to colleagues, so that they can access your work.   

You are required to write a Lab Book in **markdown** for this module, which will count as part of your evaluation. If you want, you can start a GITHub repository for the course and write your lab book there.  

**[More details on your lab notebook](https://mgxlab.github.io/Viromics2024/labbook/index.html)**


## Access to Draco

[Draco](https://wiki.uni-jena.de/pages/viewpage.action?pageId=22453002) is a high-performance cluster created and maintained by the Universit√§tsrechenzentrum. It is [available for members of Thuringian Universities](http://sternb.gitpages.tpi.uni-jena.de/draco-101-2023-01/#5).  
To log in, you can use [ssh](http://sternb.gitpages.tpi.uni-jena.de/draco-101-2023-01/#15): 

```bash
ssh <fsuid>@login1.draco.uni-jena.de
```

**[More details on access to Draco](https://mgxlab.github.io/Viromics2024/draco/index.html)**
