---
title: Getting to Know Your Sequence Data
teaching: 120
exercises: 0
questions: []
objectives:
- Interpret different file formats used in bioinformatics
- Utilize bash commands to look into sequence data
- Get a basic understanding of what sequence data looks like
- Work with sequence data in Python or R to create basic plots
keypoints:
- Sequence data and information about genes and genomes can come in many different formats
- The most common file formats are fasta (nucl. and amino acid), fastq, sam and bam, genbank, gff and tsv files
---
## Submitting jobs on Draco

When you login to Draco, you are on the "login node" and this is not the best place to run any programs or heavier scripts. You **MUST** be on a "compute node" to run scripts and programs

You can either request for resources from a compute node and run programs interactively or submit a job to the job scheduler, which then sends it to a compute node to complete.

[See here for what slurm architecture looks like](https://mgxlab.github.io/Viromics2024/draco/index.html)

To get resources for an interactive session on the short node for 1 hour with 10 cpus and 1G memory - maybe for testing a program or script

```bash
salloc --partition=short --time=01:00:00 --cpus-per-task=10 --mem=1G
```

To submit a job, you have to make a script `my_slurm_script.sh`  and then submit it as `sbatch my_slurm_script.sh` . Detailed information on creating these scripts, including descriptions can be found in the "extras" [here](https://mgxlab.github.io/Viromics2024/sbatch/index.html)

[Example script here](https://mgxlab.github.io/Viromics2024/sbatch/index.html#example-sbatch-script)

## Navigating your filesystem space

You can use the following to get into Draco

```bash
ssh <fsuid>@login1.draco.uni-jena.de

# OR

ssh <fsuid>@login2.draco.uni-jena.de
```

Take a look at the file structure and files. We have given you a subset of viromics data. Some helpful commands might be

```bash
# list
ls
ll

# path to working directory
pwd

# print first 10 lines
head -10
# print last 10 lines
tail -10

# open file on terminal (press 'q' to exit less)
less
more

# count number of lines
wc -l file.txt

# search for "@" inside file 
grep "@" file.fastq

# count number of sequence headers in a fasta file
grep -c ">" file.fasta

```

[See more useful commands and one-liners here](https://github.com/vmkhot/useful-scripts/blob/main/Linux%20Commands%20Cheat%20Sheet.md#linux-commands-cheat-sheet)



## Understanding bioinformatics file formats

The first video covers different file formats commonly used in bioinformatics. Namely, FASTA, FASTQ, BAM, SAM, VCF, BCF, GFF, GTF and BED files.

[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/KZ2wqKFerG0/0.jpg)](https://www.youtube.com/watch?v=KZ2wqKFerG0&ab_channel=edu-ome)

The second video covers different types of sequence files, including fasta and fastq, phred scores, files containing metadata (tsv files) and compressed files.

[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/D4WDdAbZW1Y/0.jpg)]( https://www.youtube.com/watch?v=D4WDdAbZW1Y&ab_channel=BasE.Dutilh)

> ## Exercise
>
> Describe fastq and fasta file formats
> 
> - What information is contained in sequencing files?
>     - Choose 2 (other than fasta) and describe
> - How many lines in your fastq files?
>     - How many sequences?
> - Print the first 10 and last 10 in terminal
>     - What's the difference?
> {: .source}
{: .challenge}
