---
start: True
title: "Assembly lecture"
teaching: 120
exercises: 30
questions:
objectives:
- "Watch the lecture videos and read about assembly algorithms"
keypoints:
- "de-Bruin graphs can be used to assemble genomes from reads"
---

# Assembly video lecture and self study

Watch the lectures below and write down at least 3 questions and/or discussion points about them. First, watch 
the lecture "Assembly strategies for genomics and metagenomics" by Bas Dutilh. It will give you an overview about
reference-guided and de-novo assembly of genomic sequences from sequencing data:
  
[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/mHmMbPxKmn0/0.jpg)](https://www.youtube.com/watch?v=mHmMbPxKmn0)  

Click on the image to watch the video "Sequence alignment" by Bas Dutilh. It will give you additional information
on alignments and multi-sequence alignments, which are used in most assembly strategies and generally, most 
bioinformatic analyses:
  
[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/MgdfZTA-J3o/0.jpg)](https://www.youtube.com/watch?v=MgdfZTA-J3o)


> ## Understand the lectures
> Understand the difference between assembly algorithms. What are the strengths and weaknesses of
> DBG (De-Bruijn Graph) and OLC (Overlap-Layout-Consensus) assemblies? Which would you use to determine the
> mutations of a model organism after an evolutionary experiment and which would you use to determine the
> genome sequence of an unknown organism?
{: .discussion}

# Supplementary reference 

The book "Computational Biology: Genomes, Networks, Evolution MIT course 6.047/6.878" by Prof. Manolis Kellis 
is part of a course on Computational Biology and contains several topics that are relevant for Bioinformatics: 
https://ocw.mit.edu/ans7870/6/6.047/f15/MIT6_047F15_Compiled.pdf 

Read the following sections and summarize their key points in your lab book:  

- "5.2 Genome Assembly I: Overlap-Layout-Consensus Approach" and "5.3 Genome Assembly II: String graph methods" (pages 93 to 102)

