---
start: True
title: "Assembly lecture"
teaching: 60
exercises: 30
questions:
objectives:
- "Watch the lecture videos and read about assembly algorithms"
keypoints:
- "Sequence assembly can be used to assemble genomes from reads"
- "Metagenome assembly generally yields shorter contigs than genome assembly"
---

{% include base_path.html %}
<p align="center">
    <a href="{{ site.carpentries_site }}"><img src="{{ relative_root_path }}/assets/img/logo_day2.png" alt="Viromics workflow" width="1600" /></a>
</p>

*Sequence assembly* is the reconstruction of long contiguous sequences (called *contigs* or *scaffolds*, see video below) from short sequencing reads. Before 2014, a common approach in metagenomics was to compare the short sequencing reads to the genomes of known organisms in the database (and some studies today still take this approach). However, this only works if the organisms in the database are closely related to the ones in the metagenomic sample. Recall that most of the sequences in a metavirome are unknown ("viral dark matter"), meaning that they yield no matches when compared to the reference database. Because of this, we need database-independent approaches to reconstruct new viral sequences. As sequencing technology and bioinformatic tools improved, sequence assembly enabled the recovery of longer sequences from metagenomic data. Having a longer sequence means having more information to classify it, so using metagenome assembly helps to characterize complex communities.

## Video on sequence assembly

Watch the lecture video "Assembly strategies for genomics and metagenomics". It will introduce reference-guided and de-novo assembly of genomic and metagenomic sequences (56 minutes):

> ## Discussion
> Watch the lecture video below and write down at least 3 questions and/or discussion points about it.
> {: .source}
{: .challenge}

- Click on the image to see lecture video "Assembly strategies for genomics and metagenomics" by Prof Bas E. Dutilh (56 minutes):

[![Lecture video "Assembly strategies for genomics and metagenomics" by Prof. Bas E. Dutilh](https://img.youtube.com/vi/mHmMbPxKmn0/0.jpg)](https://www.youtube.com/watch?v=mHmMbPxKmn0)  

> ## Questions
> 1. Would you use DBG (De-Bruijn Graph) or OLC (Overlap-Layout-Consensus) to assemble a dataset consisting of one billion short sequencing reads?
> 2. What are the strengths and weaknesses of reference-guided assembly and *de novo* assembly?
> 3. Would you use reference-guided or *de novo* assembly to assemble the genome of a model organism to discover mutations that occurred during an evolutionary experiment?
> 4. Would you use reference-guided or *de novo* assembly to determine the genome sequence of an unknown organism?
> 5. Why does metagenome assembly generally yield shorter contigs than genome assembly?
{: .challenge}

Additional reading: [Computational Biology: Genomes, Networks, Evolution MIT course 6.047/6.878 (Prof. Manolis Kellis)](https://ocw.mit.edu/ans7870/6/6.047/f15/MIT6_047F15_Compiled.pdf). This book is part of a course on Computational Biology and contains several topics that are relevant for Bioinformatics.

> ## Read and summarize
> Read the following sections and summarize shortly their key points in your lab book:
> - "5.2 Genome Assembly I: Overlap-Layout-Consensus Approach" and "5.3 Genome Assembly II: String graph methods" (pages 93 to 102).
{: .challenge}
