---
title: "Visualizing the assembly"
teaching: 0
exercises: 60
questions:
objectives:
- "Understand the topology of the de-Bruijn graph"
- "Understand how the presence of similar species in the sample affects the assembly"
keypoints:
- "Bandage can visualize the de-Bruijn graph"
- "JBrowse2 can visualize genomic data like alignments and coverage"
---

# Visualizing the assembly

As a homework, you can choose one of the following two topics. Pick what sounds more
interesting to you and if you find the time during the rest of the course, discuss your 
results with someone in the course who picked the other topic. (Bandage is difficult to
run on Windows)

## Paths in the de-Bruijn graph

We will use the [Bandage](https://github.com/rrwick/Bandage) tool to visualize
the graph. In the [releases](https://github.com/rrwick/Bandage/releases/) section,
follow instructions to download the most appropriate version. If you are in attending
the workshop live, download `Bandage_Ubuntu-x86-64_v0.9.0_AppImage.zip`).
To run it, unzip the file, open a new terminal tab (Ctrl + Shift + t), activate the conda
environment for today, and call Bandage from the terminal like this:

~~~
# run Bandage
$ ./Bandage_Ubuntu-x86-64_v0.9.0.AppImage
~~~
{: .language-bash}

In _File > Load_graph_, navigate to any of the per sample assemblies and load
`assembly_graph.fastg` for the k-mer size 21. Then click _Draw graph_ to see the
graph. Note well this graph has been already compacted by collapsing those nodes
that form linear, unbranched paths. Nodes in the graph are called edge_N 
(confusing name, with n being an integer) and they often correspond to final 
contigs in your assembly.

> ## The repeat graph and overlaps between contigs
> Open the file `assembly_info.txt` which corresponds to the graph you are looking at.
> The N in the names of the nodes as displayed by Bandage corresponds to the number
> assinged to continuous paths in the de-Bruijn graph by flye. The column called
> "graph_path" in the file `assembly_info.txt` holds this information for all contigs.
> Can you explain the structure of the information presented in this column by relating
> it to the visualization in Bandage (What does an asterisk * mean? What do multiple
> occurences of the same number mean)? Pick two components of the visualized de-Bruijn
> graph and explain their topology and information content.
{: .discussion}

## Effect of related species

Often in metagenomic samples, a number of strains for a given species are present,
and this is particularly evident with viral communities that typically contain an
abundance of haplotypes (or quasispecies). Because of the high amount of homologous
regions between these strains, the assembly graph is complex as multiple genomes
occupy much of the same kmer space. The convergence-divergence structure in the graph
generated by these homologous regions make traversing the graph more complex, and
mistakes at this point can lead to chimeric contigs containing sequence from more
than one strain. To visualize this effect, we will align the reads in our samples
back to the contigs that were assembled and use [jbrowse2](https://jbrowse.org/jb2/download/)
to visualize the differences bettween reads and assembled contigs. Under Linux or
Mac, you can start the AppImage simply by typing

~~~
# run JBrowse2
$ ./jbrowse-desktop-v2.13.1-linux.AppImage
~~~
{: .language-bash}

from within the corresponding folder. After starting JBrowse2, select 
`open sequence file(s)`: 

{% include base_path.html %}
<p align="center">
    <a href="{{ site.carpentries_site }}"><img src="{{ relative_root_path }}/assets/img/jbrowse2_step1.png" alt="Jbrowse2 - new session" width="800" /></a>
</p>

Assign a name to your assembly, select FastaAdapter under 
`Type` and then `choose file` to navigate to your assembly file and `submit`:

<p align="center">
    <a href="{{ site.carpentries_site }}"><img src="{{ relative_root_path }}/assets/img/jbrowse2_step2.png" alt="Jbrowse2 - open assembly" width="600" /></a>
</p>

Click `launch view` with `linear genome view` selected, then `show all regions in assembly` 
and "open track selector", click the plus sign in the lower right corner and select 
`add track`. Under `main file`, select `File` and then `choose file` and navigate to 
the alignments you computed for the first of our three samples (e.g., "barcode62.bam"). 
After that, repeat the same under `index file` and navigate to the index file of the 
alignments (e.g., "barcode62.bam.csi"). Click `Next` and make sure, the `IndexedBamAdapter`
is selected and confirm.

<p align="center">
    <a href="{{ site.carpentries_site }}"><img src="{{ relative_root_path }}/assets/img/jbrowse2_step3.png" alt="Jbrowse2 - add track" width="700" /></a>
</p>

Repeat this process for the other two alignment files and their adapter files. Now, you 
can inspect single reads aligned to your assembly. Try to familiarize yourself with the 
interface. You can search for a specific contig by typing its name into the search bar 
at the top of the interface. 

> ## Find chimeric contigs
> Try to look into contigs with a high coverage and find some wich have reads mapped to
> them in all three samples. Can you identify multiple virus strains visually? Make a
> screenshot and explain what you see and how this fits into what you learned about
> assemblies with de-Bruijn graphs.
{: .discussion}
