---
start: True
title: "Assembly Lecture"
teaching: 150
exercises: 30
questions:
objectives:
- "Watch the lecture videos and read about assembly algorithms"
keypoints:
- "Sequence assembly can be used to assemble genomes from reads"
- "Metagenome assembly generally yields shorter contigs than genome assembly"
---

## Video on Sequence Assembly

First, watch the lecture video "Assembly strategies for genomics and metagenomics". It will introduce reference-guided and de-novo assembly of genomic and metagenomic sequences (56 minutes):

> ## Discussion
> Watch the lecture video below and write down at least 3 questions and/or discussion points about it.
> {: .source}
{: .challenge}

- Click on the image to see lecture video "Assembly strategies for genomics and metagenomics" by Prof Bas E. Dutilh (56 minutes):

[![Lecture video "Assembly strategies for genomics and metagenomics" by Prof. Bas E. Dutilh](https://img.youtube.com/vi/mHmMbPxKmn0/0.jpg)](https://www.youtube.com/watch?v=mHmMbPxKmn0)  

> ## Questions
> - Would you use DBG (De-Bruijn Graph) or OLC (Overlap-Layout-Consensus) to assemble a dataset consisting of one billion short sequencing reads?
> - What are the strengths and weaknesses of reference-guided assembly and *de novo* assembly?
> - Would you use reference-guided or *de novo* assembly to assemble the genome of a model organism to discover mutations that occurred during an evolutionary experiment?
> - Would you use reference-guided or *de novo* assembly to determine the genome sequence of an unknown organism?
> - Why does metagenome assembly generally yield shorter contigs than genome assembly?
{: .challenge}

Additional reading: [Computational Biology: Genomes, Networks, Evolution MIT course 6.047/6.878 (Prof. Manolis Kellis)](https://ocw.mit.edu/ans7870/6/6.047/f15/MIT6_047F15_Compiled.pdf). This book is part of a course on Computational Biology and contains several topics that are relevant for Bioinformatics.

> ## Read and summarize
> Read the following sections and summarize shortly their key points in your lab book:
> - "5.2 Genome Assembly I: Overlap-Layout-Consensus Approach" and "5.3 Genome Assembly II: String graph methods" (pages 93 to 102).
{: .challenge}
