---
title: Visualizing distributions
teaching: 180
exercises: 0
questions: []
objectives:
- plot pairwise groups of measurements and determine if they are different
keypoints:
- matplotlib and pyplot provide multiple tools for the visualization of data points
---

## Differences between assemblies.

{% include base_path.html %}
<p align="center">
    <a href="{{ site.carpentries_site }}"><img src="{{ relative_root_path }}/assets/img/violin_vs_boxplot.png" alt="Compare assemblies" width="700" /></a>
</p>

Today, you looked at the assembly of the bacterial fraction of organisms in our experiment. 
It contains much longer contigs than the assembly of the viral fraction, but also generally 
more of them. As a Homework, try to visualize the distribution of the lengths of the contigs. 
Matplotlib provides many tools to [visualize data and their distribution](https://matplotlib.org/stable/plot_types/index.html).
Try to use two or more of these (e.g. histogram, violinplot or boxplot) and adjust the plots
in a way that it confers the differences in the distributions well. The bacterial assembly
contains some very long fragments (outliers), try to find a way to deal with those.  

> ## python script for converting tabular annotations into the GFF format
> ```python
> import os, sys
> from Bio import SeqIO
> import matplotlib.pyplot as plt
> 
> def main():
> 
>     assembly1_filename = os.path.abspath(sys.argv[1])
>     assert assembly1_filename.endswith(".fasta")
>     
>     assembly2_filename = os.path.abspath(sys.argv[2])
>     assert assembly2_filename.endswith(".fasta")
>     
>     out_filename = os.path.abspath(sys.argv[3])
>     assert out_filename.endswith(".png")
> 
>     with open(assembly1_filename) as handle:
>         lengths_assembly1 = [len(record.seq) for record in SeqIO.parse(handle, "fasta")]
>         
>     with open(assembly2_filename) as handle:
>         lengths_assembly2 = [len(record.seq) for record in SeqIO.parse(handle, "fasta")]
> 
>     fig, axs = plt.subplots(nrows=1, ncols=2, figsize=(9, 4))
>     
>     axs[0].violinplot([lengths_assembly1, lengths_assembly2], showextrema=False)
>     axs[0].set_xticks([1,2], ["phages", "bacteria"])
>     axs[0].set_ylim([0,100000])
>     axs[0].set_title('Violin plot')
>     
>     axs[1].boxplot([lengths_assembly1, lengths_assembly2], showfliers=False)
>     axs[1].set_xticks([1,2], ["phages", "bacteria"])
>     axs[1].set_title('Box plot')
>     
>     fig.savefig(out_filename, dpi=200)
>     
> 
> if __name__ == "__main__":
>     main()
>```
> {: .source}
{: .solution}

> ## Plot the difference between two assemblies
> Plot the distribution of the lengths of both, the bacterial and the viral assemblies using matplotlib. Choose at least 2 of the following:
> - histogram
> - boxplot
> - violinplot
> What are the strengts and weaknesses of your chosen methods, how do they compare in highlighting the differences between the length distributions of the two assemblies.
{: .challenge}
