---
start: True
title: "Identifying Viral Contigs II"
exercises: 180
questions:
- "Before running Jaeger: where do you expect the most hits (virome or metagenome)? Why?"
- "After running Jaeger: do the results corroborate your expectations?"
- "After: why are not all contigs in the virome identified as viral contigs?"
- "After: why are viral contigs also identified in the metagenome?"
- "After/optional: how many viral contigs are there in each assembly?"
- "After/optional: take the longest contig from each dataset and blast it (link below). What are the top hits?"
- "After/optional: what are the size differences between viral and non-viral contigs (you investigate this using visualization)?"
objectives:
- "does this work?"
keypoints:
---

- Link for BLAST: https://blast.ncbi.nlm.nih.gov/Blast.cgi

In this experimental part, you will use [Jaeger](https://github.com/Yasas1994/Jaeger) to identify viral contigs. You will run Jaeger on the viromes you assembled and also on metagenomes, which were produced by Dr Xiu Jia for the community coalescence project. You will do that to compare Jaeger's results when running it for viromes and metagenomes. Answer the questions before and after running Jaeger.   

First, log in to draco, allocate a node and activate Jaeger's conda environment. Then, run it for viromes and metagenomes. Expected run time for viromes: 30' and for metagenomes: 1h 30'. Read and interpret the output following Jaeger's [tutorial](https://github.com/Yasas1994/Jaeger?tab=readme-ov-file#what-is-in-the-output).

```bash
salloc -p gpu
source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh && conda activate jaeger_v1.31.0 && Jaeger -h
#Metagenomes
Jaeger -i /work/groups/VEO/shared_data/veo_students/metagenome_XJ/bacterial_assembly_q15.fasta -o bacterial_assembly_q15
#Viromes
Jaeger -i input_file.fasta -o output_dir
head output_dir/output_dir.tsv
```

While Jaeger runs, you can already prepare for the next steps using the example below of Jaeger's output:

```
contig_id       length  #num_prok_windows       #num_vir_windows        #num_fun_windows        #num_arch_windows       prediction      bac_score       vir_score       fun_score       arch_score      window_summary
contig_1        12924   0       6       0       0       Phage   0.2848  4.7415  -6.2909 -4.4449 6V
contig_10       11502   0       5       0       0       Phage   0.8859  4.0839  -6.4333 -4.4140 5V
contig_100      2167    0       1       0       0       Phage   2.9913  3.3828  -6.7681 -5.9237 1V
contig_1000     21423   0       9       1       0       Phage   -0.3513 4.2560  -4.1227 -4.5463 9V1n
contig_10001    11520   0       2       3       0       Phage   -1.9324 0.8609  0.6485  -5.2195 2V3n
contig_10010    7109    2       1       0       0       Non-phage       2.3482  0.8580  -2.1243 -5.15051n1V1n
```

Use Jaeger's predictions of column 7 to separate phage from non-phage contigs. If you are an experienced programmer, write (a) script(s) to do that. If you are not, copy the two python codes below to two files in draco and run them.   

```
conda deactivate
conda activate /home/no58rok/tools/miniconda3/envs/bacterial_phenotypes
vim select_contigs.py
python3 select_contigs.py > phage_contig
python3 filter_FASTA2.py > phageContigs.fasta
```

Script select_contigs.py:  

```python 
import pandas as pd

#Get name of path and file
filename = 'viral_assembly_all/tmp.tsv'
#filename = 'path'

#Read the file into a pandas DataFrame
df = pd.read_csv(filename, sep='\s+')
#print(df.head())

#Filter rows where the seventh column is "Phage" and print the contig ID
for i in range(len(df)):

    #Get prediction and ID
    prediction = df.loc[i, 'prediction']
    contig_id = df.loc[i, 'contig_id']

    #If prediction is Phage then print the contig ID
    if(prediction == 'Phage'):
        print(contig_id)
```
Script filter_FASTA2.py:  

```python
# Define the input and output file names
#input_fasta = '/home/groups/VEO/shared_data/xiu/viral_assembly_all.fasta'
input_fasta = 'tmp.fasta'
input_list = 'phage_contigs.txt'

# Read the list of contigs from the text file
with open(input_list, 'r') as file:
    lst = [line.strip() for line in file]

def read_fasta_to_dict(fasta_file):
    sequences = {}
    with open(fasta_file, 'r') as infile:
        current_header = None
        current_sequence = []
        for line in infile:
            line = line.strip()
            if line.startswith('>'):
                if current_header:
                    sequences[current_header] = ''.join(current_sequence)
                current_header = line[1:]  # Remove the '>'
                current_sequence = []
            else:
                current_sequence.append(line)
        if current_header:
            sequences[current_header] = ''.join(current_sequence)
    return sequences

# Read the FASTA file and store the sequences in a dictionary
sequences_dict = read_fasta_to_dict(input_fasta)

# Print the dictionary if the contig ID is in the FASTA header
for header, sequence in sequences_dict.items():
    if any(contig in header for contig in lst):
        print(f">{header}\n{sequence}")
```

After selecting phage contigs and saving it to a new FASTA file, assess the genome completeness of the contigs and filter only high quality ones.      

Optional tasks:

- Plot Jaegar scores per genome length



        
- How would you rank Jaeger compared to other virus detection tools? (After running tool)

Check completeness with CheckV
