---
start: True
title: "Identifying Viral Contigs II"
exercises: 180
questions:
- "Before running Jaeger: where do you expect the most hits (virome or metagenome)? Why?"
objectives:
- "does this work?"
keypoints:
---

# Identify viral contigs

In this experimental part, you will use [Jaeger](https://github.com/Yasas1994/Jaeger) to identify viral contigs. You will run Jaeger on the viromes you assembled and also on metagenomes, which were produced by Dr Xiu Jia for the community coalescence project. You will do that to compare Jaeger's results when running it for viromes and metagenomes. Answer the questions before and after running Jaeger.   

To run Jaeger, first activate Jaeger's conda environment. Then, run it for viromes and metagenomes. Read and interpret the output following Jaeger's [tutorial](https://github.com/Yasas1994/Jaeger?tab=readme-ov-file#what-is-in-the-output).

```bash
source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh && conda activate jaeger_v1.31.0 && Jaeger -h
Jaeger -i input_file.fasta -o output_dir
head output_dir/output_dir.tsv
```

After running Jaeger, you will assess genome completeness and filter only high quality ones to use for subsequent analysis.  


After running Jaegar
Questions to think about
      - Take the longest contig from each dataset and blast it
      - Plot from jaegar score per genome length?
      - How many viral contigs are there in each assembly?
      - Why are not all contigs in the virome identified as viral contigs?
      - Why are viral contigs also identified in the metagenome?
      - size differences between viral and non-viral contigs
        
    - How would you rank Jaeger compared to other virus detection tools? (After running tool)

Check completeness with CheckV
