---
start: True
title: "Identifying Viral Contigs II"
exercises: 180
questions:
objectives:
- "Identify phage contigs in assembled viromes and metagenomes"
- "Interpret Jaeger's output"
- "Select viral contigs for further analysis"
- "Assess assembly completeness"
- "Understand differences between viromes and metagenomes"
keypoints:
- ""
---

In this experimental part, you will identify viruses in your assembled contigs. In the morning, you learned about different virus identification tools with a benchmarking paper. Today, you will use a tool that was developed after that benchmarking: [Jaeger](https://github.com/Yasas1994/Jaeger). Jaeger was developed by the VEO-MGX groups. You will run it on the viromes you assembled and also on metagenomes that were produced for the same project (community coalescence). You will do that to compare Jaeger's results for viromes and metagenomes. Answer the questions before and after running Jaeger.   

# Identify Viral Contigs

> ## discussion
>
> Viromes and metagenomes are prepared differently in the wet lab. Each has their purpose. Where do you expect the most hits from Jaeger (virome or metagenome) and why? 
> 
> Why is it important to use tools like Jaeger on viromes?  
> 
> {: .source}
{: .discussion}

Location of metagenome assemblies: ```/work/groups/VEO/shared_data/veo_students/metagenome_XJ/bacterial_assembly_q15.fasta```

> ## exercise
>
> Run Jaeger for viromes and metagenomes
> - read and interpret the output following Jaeger's [tutorial](https://github.com/Yasas1994/Jaeger?tab=readme-ov-file#what-is-in-the-output).
> 
>```bash
> source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh && conda activate jaeger_v1.31.0 && Jaeger -h
> Jaeger -i ? -o ?
>```
>
> - inspect the output
>   
>```bash
> head <output>
>```
> {: .source}
{: .challenge}
> 

After running Jaeger, answer the questions below.

> ## discussion
>
> do the results corroborate your expectations?
> why are not all contigs in the virome identified as viral contigs?
> why are viral contigs also identified in the metagenome?
>
> optional: how many viral contigs are there in each assembly?
> optional: take the longest contig from each dataset and BLAST it (link below). What are the top hits, are they expected?
> optional: what are the size differences between viral and non-viral contigs (you investigate this using visualization)?
> optional: how would you rank Jaeger compared to other virus detection tools mentioned in the benchmark paper? For this you could run the virome with other tools and compare the outputs.
> 
> {: .source}
{: .discussion}

# Filter Contigs

Use Jaeger's predictions of column 7 to separate phage from non-phage contigs. If you are an experienced programmer, write (a) script(s) for that. If you are not, use the solutions below.   

> ## bash commands for conda and running scripts
> ```bash
> conda deactivate
> conda activate /home/no58rok/tools/miniconda3/envs/bacterial_phenotypes
> vim select_contigs.py
> python3 select_contigs.py > phage_contig
> python3 filter_FASTA2.py > phageContigs.fasta
>```
> {: .source}
{: .solution}

> ## script for selecting viral contigs  
> ```python 
> import pandas as pd
> 
> #Get name of path and file
> filename = 'viral_assembly_all/tmp.tsv'
> #filename = 'path'
> 
> #Read the file into a pandas DataFrame
> df = pd.read_csv(filename, sep='\s+')
> #print(df.head())
> 
> #Filter rows where the seventh column is "Phage" and print the contig ID
> for i in range(len(df)):
> 
>     #Get prediction and ID
>     prediction = df.loc[i, 'prediction']
>     contig_id = df.loc[i, 'contig_id']
> 
>     #If prediction is Phage then print the contig ID
>     if(prediction == 'Phage'):
>         print(contig_id)
>```
> {: .source}
{: .solution} 

> ## script for filtering FASTA file  
>
>```python
> # Define the input and output file names
> #input_fasta = '/home/groups/VEO/shared_data/xiu/viral_assembly_all.fasta'
> input_fasta = 'tmp.fasta'
> input_list = 'phage_contigs.txt'
> 
> # Read the list of contigs from the text file
> with open(input_list, 'r') as file:
>     lst = [line.strip() for line in file]
> 
> def read_fasta_to_dict(fasta_file):
>     sequences = {}
>     with open(fasta_file, 'r') as infile:
>         current_header = None
>         current_sequence = []
>         for line in infile:
>             line = line.strip()
>             if line.startswith('>'):
>                 if current_header:
>                     sequences[current_header] = ''.join(current_sequence)
>                 current_header = line[1:]  # Remove the '>'
>                 current_sequence = []
>             else:
>                 current_sequence.append(line)
>         if current_header:
>             sequences[current_header] = ''.join(current_sequence)
>     return sequences
> 
> # Read the FASTA file and store the sequences in a dictionary
> sequences_dict = read_fasta_to_dict(input_fasta)
> 
> # Print the dictionary if the contig ID is in the FASTA header
> for header, sequence in sequences_dict.items():
>     if any(contig in header for contig in lst):
>         print(f">{header}\n{sequence}")
>```
> {: .source}
{: .solution} 

# Assess Completeness

After selecting phage contigs and saving them to (a) new FASTA file(s), assess the genome completeness of the contigs. For this you will use [CheckV](https://bitbucket.org/berkeleylab/checkv/src/master/).     

> ## exercise
>
> Run CheckV
> - read and interpret the output following CheckV's webpage.
> 
>```bash
> source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh && conda activate checkv_v1.0.1 && checkv -h
> checkv completeness ? -d /work/groups/VEO/databases/checkv/v1.5 ?
>```
>
> - inspect the output
>   
>```bash
> head <output>
>```
>
> The following part is optional (you could use the same logic as for filtering phage contigs)
> - keep only the contigs with enough completeness (0.5 is the usual, but you could play around with other thresholds)    
>
> {: .source}
{: .challenge}
>    

# Resources

- [Link for BLAST](https://blast.ncbi.nlm.nih.gov/Blast.cgi)
