---
start: True
title: "Identifying Viral Contigs II"
exercises: 180
questions:
- "Where do you expect the most hits? (Before running tool)"
objectives:
- "does this work?"
keypoints:
---

# Identify viral contigs

In this experimental part, you will identify viral contigs, assess genome completeness and filter only high quality ones to use for subsequent analysis.  

Before running Jaeger  
- (virome and metagenome (pre-assembled) should be in dir /path/to/dir
- virome (small size fraction) and metagenome (larger size fraction)

Activate [Jaeger](https://github.com/Yasas1994/Jaeger)'s conda environment to be able to use it, as below. Then, run Jaeger for your assemblies. Read and interprete the output following Jaeger's [tutorial](https://github.com/Yasas1994/Jaeger?tab=readme-ov-file#what-is-in-the-output).

```bash
source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh && conda activate jaeger_v1.31.0 && Jaeger -h
Jaeger -i input_file.fasta -o output_dir
more output.file
```



After running Jaegar
Questions to think about
      - Take the longest contig from each dataset and blast it
      - Plot from jaegar score per genome length?
      - How many viral contigs are there in each assembly?
      - Why are not all contigs in the virome identified as viral contigs?
      - Why are viral contigs also identified in the metagenome?
      - size differences between viral and non-viral contigs
        
    - How would you rank Jaeger compared to other virus detection tools? (After running tool)

Check completeness with CheckV
