---
title: "Gene Finding II"
teaching: 
exercises: 180
questions:
objectives:
- "Predict ORFs in viral contigs with Phanotate"
- "Compare gene annotations of Phanotate and CheckV (prodigal-gv)"
keypoints:
---

# Gene Finding

After assembling virome contigs and filtering the phage ones, you will work with genome annotation. You have actually already done an annotation to your viral contigs when you ran [CheckV](https://bitbucket.org/berkeleylab/checkv/src/master/). In its internal pipeline, CheckV uses [prodigal-gv](https://github.com/apcamargo/prodigal-gv). To complement prodigal-gv, you will do a second genome annotation with [Phanotate](https://github.com/deprekate/PHANOTATE). Afterwards you will compare the two annotations. To start, check the webpages of the two tools and answer the questions below.  

> ## discussion
>
> Describe the file format that was produced by CheckV/prodigal-gv and show its header with bash (```head <file>```).
> 
> What file formats can Phanotate produce? Which one is most similar to the one produced by prodigal-gv/CheckV?
>
> Describe the difference between Phanotate's algorithm and prodigal-gv's.
>
> Do you expect to find more hits with prodigal-gv or with Phanotate? Why?
> 
> {: .source}
{: .discussion}

Now, we ask you to produce two different outputs with Phanotate: one GFF containing the ORF coordinates and one FASTA containing the corresponding aminoacid sequences. Create sbatch scripts for running Phanotate. Note that the tool will require almost 6 GB memory, so make sure you adapt your sbatch file.

> ## exercise
>
> Run Phanotate for your viral contigs to produce a GFF file and a FASTA file:
> 
>```bash
> source /home/groups/VEO/tools/phanotate/v1.5.0/myenv/bin/activate
> phanotate.py <input> -o <output.gff> -f gff
> phanotate.py <input> -o <output.faa> -f faa
>```
>
> Inspect the output and interpret it following the tool's webpage
>   
>```bash
> head <output>
>```
> {: .source}
{: .challenge}

After running the tool, choose one contig to work with as an example. This way you can get more in depth in subsequent analysis. Choose a contig with 100% completeness and less than 5% contamination (these measures have already been calculated in the previous lecture [Identifying Viral Contigs II]([Assembly and cross-assembly](https://mgxlab.github.io/Viromics2024/1.2.2_assembly_and_cross-assembly/index.html))). Select Phanotate's predictions for that specific contig_id (gff and fasta) and save them to new files. Write a code for that if possible. If not, use the solutions below. Similarly, save the gff annotations of CheckV's prodigal-gv to a new file. We will need it to compare the annotations of Phanotate and prodigal-gv.        

> ## bash command for sorting contigs by completeness
> ```bash
> cat <file> | sort -k 10 -nr | head
>```
> {: .source}
{: .solution}

> ## bash command for saving contig files
> ```bash
> grep -P '<contig_id>\t' <output_Phanotate> > <contig_id_Phanotate.gff>
> grep -P '<contig_id>\t' <output_prodigal-gv> > <contig_id_prodigal-gv.gff>
>```
> {: .source}
{: .solution}

Next, answer the following questions in your lab book. If you are an experienced programmer, develop scripts locally in your laptop first and afterwards run it in draco using sbatch scripts. If you are not an experienced programmer, you could use the codes provided below.  

> ## discussion
> 
> What does the output file look like for your data?
> 
> How many ORFs did the tool predict per contig?
>
> What is the longest ORF per contig?
>
> Did you find any overlapping ORFs in a contig?
> 
> {: .source}
{: .discussion}

\<code to get number of orfs per contig\>
```grep -v '#' 00_phanotate/phage_contigs.gff | cut -f1 | uniq -c```

# Tools comparison

Phanotate has a different internal approach than CheckV's [prodigal-gv](https://github.com/apcamargo/prodigal-gv). While Phanotate is more broad and predicts possible ORFs, prodigal-gv predicts genes. So, we expect to see differences. It will be interesting to compare the outputs of the two tools. 

> ## discussion
> 
> How many genes did prodigal-gv predict?
>
> Offer an interpretation to the difference between results of the two tools.
> 
> {: .source}
{: .discussion}

Now that we compared the number of predictions, let's go to more details and compare the coordinates. For that we will use BEDtools intersectBed. 


> ## bash command for getting number of contigs
> ```bash
>  
/home/groups/VEO/tools/bedtools2/v2.31.0/bin/intersectBed -a <input1> -b <input2> -f 1 -r
>```
> {: .source}

> ## discussion
> 
> How many genes did prodigal-gv predict?
>
> How many overlapping elements between the two tools?
>
> How many unique elements per tool?
> 
> {: .source}
{: .discussion}
