---
start: True
title: "Genome annotation: theory"
teaching: 
exercises: 180
questions:
objectives:
- "Understand what is genome annotation"
- "Investigate the features of phages and how it impacts ORF annotation"
- "Investigate the Phanotate algorithm"
keypoints:
- "Genome annotation gives meaning to genomic sequences"
- "ORFs can be predicted from start and stop codons in the genomic sequences"
- "Phages have different genomic features than prokaryotes, which influences the design of algorithms"
- "Tools like Phanotate are very useful to process a large amount of contigs. However, no tool is perfect, so a critical interpretation of the results is important"
---

Once we have an assembled genome sequence (or even a contig that represents a fragment of a genome), we want to know  what kind of organism the sequence belongs to (**taxonomic annotation**), what genes it encodes (**gene annotation** or gene calling), and the functions of those genes (**functional annotation**). Ideally, gene calling uses a good gene model that is tailored to the organism of study, but to determine the organism we need to know its genes. So it can be an iterative process. In our case, there are pretty good standard gene models available for bacteria and for phages, which we will use.

Phage genomes have specific features that are important to keep in mind. The lecture video by Prof Robert Edwards explains these, and how they were taken into account when developing a specialized phage gene caller. Watch the video to answer the questions below. The links of [Prodigal](https://github.com/hyattpd/Prodigal) and [Phanotate](https://academic.oup.com/bioinformatics/article/35/22/4537/5480131) are also useful for answering the questions. 

- Click on the image to see the lecture video "Phage Genomics" by Prof Rob Edwards (29 minutes):
  
[![lecture video "Phage Genomics" by Prof Rob Edwards](https://img.youtube.com/vi/ecJ1DqVvuFE/0.jpg)](https://www.youtube.com/watch?v=ecJ1DqVvuFE&pp=ygUJcGhhbm90YXRl)  

As additional material, the video of [Dr Katelyn McNair](https://www.youtube.com/watch?v=gvnPsA1S6GY&pp=ygUJcGhhbm90YXRl) contains explanations of the algorithm of Phanotate. Topics are similar to the video of Dr Robert Edwards, but more detailed regarding rules and scoring.

- Click on the image to see the lecture video "Phanotate: Gene finding in phages" by Dr Katelyn McNair (67 minutes):

[![lecture video "Phanotate: Gene finding in phages" by Dr Katelyn McNair](https://img.youtube.com/vi/gvnPsA1S6GY/0.jpg)](https://www.youtube.com/watch?v=gvnPsA1S6GY&pp=ygUJcGhhbm90YXRl)

The video of [Dr Evelien Adriaenssens](https://www.youtube.com/watch?v=wO1w1Z1Or1w&pp=ygUJcGhhbm90YXRl) contains interesting information about the structure of the phage genomes and how to best annotate them.

- Click on the image to see the lecture video "Basics of phage genome annotation & classification: getting started" by Dr Evelien Adriaenssens (68 minutes, watch from 12:06 to 29:57):
  
[![lecture video "Basics of phage genome annotation & classification: getting started" by Dr. Evelien Adriaenssens](https://img.youtube.com/vi/wO1w1Z1Or1w/0.jpg)](https://www.youtube.com/watch?v=wO1w1Z1Or1w)

Answer the following questions in your lab book:  

> ## Questions
> 
> 1. What is genome annotation?
> 
> 2. What are ORFs and why are they important in annotating genes?
>
> 3. Explain the Phanotate algorithm.
> 
> 4. Which features of phages should be considered when developing a phage gene annotation tool?
>
> 5. Would the tool Prodigal be adequate for annotating genes in phage genomes? Why?
> {: .source}
{: .discussion}

Below are optional questions, which you should only tackle if you fully understood the material above:

> ## Additional questions (optional)
> 
> 6. How did the developers of Phanotate compare their predictions with other tools? Do you think this is an adequate benchmark?
>
> 7. What are the main classes of proteins of phage lambda?
>
> 8. How is the tool Phanns different from Phanotate?
> {: .source}
{: .discussion}
