---
title: "Gene Finding II"
teaching: 
exercises: 180
questions:
objectives:
- "Predict ORFs in viral contigs"
keypoints:
---

# Gene Finding

After assembling virome contigs and filtering the phage ones, you will annotate them using [Phanotate](https://github.com/deprekate/PHANOTATE). Check their GITHub page to know what to expect when running it.   

> ## discussion
> 
> What file formats can Phanotate produce? Describe them.
> 
> {: .source}
{: .discussion}

We ask you to produce two different outputs with Phanotate, one GFF and one FASTA containing the protein sequences. Create two different sbatch scripts and submit the two jobs in parallel.

> ## exercise
>
> Run Phanotate for your viral contigs to produce a GFF file and a FASTA file:
> 
>```bash
> source /home/groups/VEO/tools/phanotate/v1.5.0/myenv/bin/activate
> phanotate.py ? -o output.gff -f gff
> phanotate.py ? -o output.faa -f faa
>```
>
> inspect and interpret the output following the tool's webpage
>   
>```bash
> head <output>
>```
> {: .source}
{: .challenge}

After running the tool, answer the following questions in your lab book. Note that Phanotate outputs files in different formats (check [Phanotate](https://github.com/deprekate/PHANOTATE)).  The genebank file contains a human readable format of the annotation, the GFF file a tabular format. Inspect these files so that you are familiar with their formats. See https://en.wikipedia.org/wiki/General_feature_format for some more information on GFF.

> ## discussion
> 
> What does the output file look like?
> 
> How many ORFs did the tool predict per contig?
>
> What is the longest ORF per contig?
>
> Did you find any overlapping ORFs in a contig?
>
> Find similar contigs with [AAI clustering method](https://github.com/snayfach/MGV/tree/master/aai_cluster)
> 
> {: .source}
{: .discussion}

