---
title: "Visualizing a viral genome"
teaching: 
exercises: 60
questions:
objectives:
- "Use python to draw a map of a viral genome"
- "Visualize gene overlap and direction"
keypoints:
- "The pyCirclize package provides tools for plotting genomic data, e.g., the set of genes, in a circular layout."
---

# A circular genome map

You will use python to visualize the genes you annotated today. The 
[pyCirclize](https://github.com/moshi4/pyCirclize) package provides tools to generate
high quality plots of genomic data in a circular layout. Check the link to the github
page to get an idea of its possibilities. To use the package, you have to install it
into your virtual environment: 

~~~
# activate the environment
$ source path/to/your/py3env/bin/activate

# install the packages we need
$ pip install pyCirclize
~~~
{: .language-bash}

You can get some examples of how to use the package [here](https://moshi4.github.io/pyCirclize/circos_plot/).
You can use first example, the circos plot of the Enterobacter Phage, and adjust it to
use the annotated ORFs you created today with phannotate and the ORFs annotated by CheckV
on the second day. We do not have functional annotations of the ORFs in both cases, so 
you can skip the labelling of the ORFs. The information about the ORFs used by CheckV can
be found in the `tmp` folder in your CheckV results (`gene_features.tsv`). Think about a
way to convert the information in the TSV file into something [parsable by pyCirclize](https://github.com/moshi4/pyCirclize/tree/main/src/pycirclize/parser) 
first and then write a script for plotting the two sets of ORFs.

> ## python script for plotting ORFs with pyCirclize
> ```python
> from pycirclize import Circos
> from pycirclize.parser import Genbank, Gff
> import os
> 
> # Load Genbank file
> gbk_file = load_prokaryote_example_file("enterobacteria_phage.gbk")
> gbk = Genbank(gbk_file)
> 
> # Initialize circos sector by genome size
> circos = Circos(sectors={gbk.name: gbk.range_size})
> circos.text("Enterobacteria phage\n(NC_000902)", size=15)
> sector = circos.sectors[0]
> 
> # Outer track
> outer_track = sector.add_track((98, 100))
> outer_track.axis(fc="lightgrey")
> outer_track.xticks_by_interval(5000, label_formatter=lambda v: f"{v / 1000:.0f} Kb")
> outer_track.xticks_by_interval(1000, tick_length=1, show_label=False)
> 
> # Plot forward & reverse CDS genomic features
> cds_track = sector.add_track((90, 95))
> cds_track.genomic_features(gbk.extract_features("CDS", target_strand=1), plotstyle="arrow", fc="salmon")
> cds_track.genomic_features(gbk.extract_features("CDS", target_strand=-1), plotstyle="arrow", fc="skyblue")
> 
> out_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), "20_viral_genome")
> if not os.path.exists(out_dir): os.mkdir(out_dir)
> out_file = os.path.join(out_dir, "circos_genome.png")
> circos.savefig(out_file)
>```
> {: .source}
{: .solution}

> ## Compare the annotations
> Describe the differences between the two annotations based on the two plots you created.
> Does the visualization confer these differences well? 
{: .challenge}
