---
start: True
title: "Gene Finding I"
teaching: 120
exercises: 30
questions:
- "What is genome annotation?"
- "What are ORFs and why are they important in annotating genes?"
- "How did the developers of Phanotate compared their ORF predictions with other tools?"
- "What are the main classes of proteins of phage lambda?"
- "How is tool Phanns different from Phanotate?"
- "Which particularites of phages should be considering when developing a phage annotation tool?"
- "Would the tool prodigal be adequate for annotating genes in phage genomes? Why?"
objectives:
- "does this work?"
keypoints:
---

# Gene Finding

The process of genome annotation should be tailored to the organism of study. Phages have specificities in their genomes and evolutionary processes that are important to keep in mind. The [video of Dr Robert Edwards](https://www.youtube.com/watch?v=ecJ1DqVvuFE&pp=ygUJcGhhbm90YXRl) explains particularities about the phage genome that were taken into account when developing an annotation tool for phages. Watch it to answer the questions of this tutorial. The links of [Prodigal](https://github.com/hyattpd/Prodigal) and [Phanotate](https://academic.oup.com/bioinformatics/article/35/22/4537/5480131) are also useful for answering the questions.    

Katelyn: https://www.youtube.com/watch?v=gvnPsA1S6GY&pp=ygUJcGhhbm90YXRl
Evelien Adrianssens: https://www.youtube.com/watch?v=wO1w1Z1Or1w&pp=ygUJcGhhbm90YXRl


