---
title: "Gene Finding II"
teaching: 
exercises: 180
questions:
objectives:
- "Predict ORFs in viral contigs with Phanotate"
- "Compare gene annotations of Phanotate and CheckV (prodigal-gv)"
keypoints:
---

# Gene Finding

After assembling virome contigs and filtering the phage ones, you will find ORFs with [Phanotate](https://github.com/deprekate/PHANOTATE). Check their GITHub page to know what to expect when running it.   

> ## discussion
> 
> What file formats can Phanotate produce? Describe them.
> 
> {: .source}
{: .discussion}

We ask you to produce two different outputs with Phanotate, one GFF containing the ORF coordinates and one FASTA containing the corresponding aminoacid sequences. Create sbatch scripts for running Phanotate. Note that the tool will require almost 6 GB memory, so make sure you adapt your sbatch file.

> ## exercise
>
> Run Phanotate for your viral contigs to produce a GFF file and a FASTA file:
> 
>```bash
> source /home/groups/VEO/tools/phanotate/v1.5.0/myenv/bin/activate
> phanotate.py <input> -o output.gff -f gff
> phanotate.py <input> -o output.faa -f faa
>```
>
> inspect and interpret the output following the tool's webpage
>   
>```bash
> head <output>
>```
> {: .source}
{: .challenge}

After running the tool, choose one contig to work with as an example and to get more in depth. For that, select Phanotate's predictions for that specific contig and save it to a new file. Save the annotations of CheckV's prodigal-gv in a new file as well, since we will need it for subsequent comparisons.    

> ## bash command for selecting a contig
> ```bash
> grep -P 'contig_10\t' 00_phanotate/phage_contigs.gff > 00_phanotate/contig_21.gff
> grep -P 'contig_10\t' ../1.2_virome_assembly/21_results_assessment_checkv/crossassembly/tmp/gene_features.tsv > 00_phanotate/contig_10_prodigal-gv.gff
>```
> {: .source}
{: .solution}

Next, answer the following questions in your lab book. If you are an experienced programmer, develop scripts if needed locally and then run it in draco using sbatch scripts. If you are not an experienced programmer, you could use the codes provided below.  

> ## discussion
> 
> What does the output file look like for your data?
> 
> How many ORFs did the tool predict per contig?
>
> What is the longest ORF per contig?
>
> Did you find any overlapping ORFs in a contig?
> 
> {: .source}
{: .discussion}

\<code to get number of orfs per contig\>
```grep -v '#' 00_phanotate/phage_contigs.gff | cut -f1 | uniq -c```

# Tools comparison

Phanotate has a different internal approach than CheckV's [prodigal-gv](https://github.com/apcamargo/prodigal-gv). While Phanotate is more broad and predicts possible ORFs, prodigal-gv predicts genes. So, we expect to see differences. It will be interesting to compare the outputs of the two tools. 

> ## discussion
> 
> How many genes did prodigal-gv predict?
>
> Offer an interpretation to the difference between results of the two tools.
> 
> {: .source}
{: .discussion}

Now that we compared the number of predictions, let's go to more details and compare the coordinates. For that we will use BEDtools intersectBed. 


> ## bash command for getting number of contigs
> ```bash
>  
/home/groups/VEO/tools/bedtools2/v2.31.0/bin/intersectBed -a <input1> -b <input2> -f 1 -r
>```
> {: .source}

> ## discussion
> 
> How many genes did prodigal-gv predict?
>
> How many overlapping elements between the two tools?
>
> How many unique elements per tool?
> 
> {: .source}
{: .discussion}
