---
start: True
title: "Gene Calling and Functional Annotation II"
teaching: 
exercises: 180
questions:
objectives:
- "Perform gene calling and functional annotation on the viral contigs"
keypoints:
- Functional annotation of viral genomes can give clues about viral lifestyle and host interactions
---

## Running pharokka

For the functional annotation of our viral contigs , we will be using Pharokka. 

We first have to load the appropriate conda environment

```bash

source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh

conda activate pharokka_v1.7.1

```

The base usage for this program follows:

```bash

pharokka.py -i <fasta file> -o <output folder> -d <path/to/database_dir> -t <threads>  

```

See the [docs](https://pharokka.readthedocs.io/en/latest/run/) for the pharokka parameters

We have to add the following options:  
- To run on viral contigs from a metavirome: `-m`
- To also use PyHMMER `--meta_hmm`
- To use Phanotate for the gene prediction: `-g phanotate` 

> ## Exercise
>  
>  1. Run pharokka with the appropriate parameters using an sbatch script
>
> {: .source}
{: .challenge}

> ## sbatch script for pharokka
>```bash
>#!/bin/bash
>#SBATCH --tasks=1
>#SBATCH --cpus-per-task=10
>#SBATCH --partition=short,standard,interactive
>#SBATCH --mem=20G
>#SBATCH --time=3:00:00
>#SBATCH --job-name=pharokka
>#SBATCH --output=pharokka.slurm.out.%j
>#SBATCH --error=pharokka.slurm.err.%j
>
>source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh
>
>conda activate pharokka_v1.7.1
>
>
>#database_path /work/groups/VEO/databases/pharokka/v1.7.1/
>
>
>echo `date`
>echo 'Running meta_hmm wih phanotate mode'
>
>pharokka.py -i phage_contigs.fasta -o ./pharokka_output \
>-d /work/groups/VEO/databases/pharokka/v1.7.1/ \
>-t 10 -m --meta_hmm -f -g phanotate
>
>echo `date`
>
>```
> {: .source}
{: .solution}

## Plotting functional annotations

### Choose a pet contig (or two)

> ## Exercise - extracting contig information
>
>{:start='2'}
>  2. Choose a pet contig. Maybe your favourite contig number... but we recommend choosing one with high completeness and which is 'annotation-rich' - perhaps with many annotated genes, a crispr region, mash results (from pharokka) etc. You might want to discuss with your classmates so that not everybody picks the same one. We will continue using this "pet" contig in the future lessons also.
>      a) Write down some information about your contig. e.g. contig length, # of genes, # of annotated genes, gc%, coding density
>
> {: .source}
{: .challenge}


For the plotting we, we will still use pharokka. Pharokka's built-in `pharokka_plotter.py` uses a python library called [pyCirclize](https://github.com/moshi4/pyCirclize). You will also use this library later for the visualization homework so stay tuned!

pharokka_plotter.py requires a gff, genbank and fasta file for the viral contig. Therefore, we have to extract these from the larger pharokka.gff and pharokka.gbk outputs.

> ## Exercise - extracting contig information
>
>{:start='3'}
> 3. Extract the functional annotation information for your contig from the `pharokka.gff` and `pharokka.gbk` files and the contig nucleotide sequence from the fasta file
>
>  {: .source}
{: .challenge}

Gff and fasta files are easy to pull information from so we can just `grep`  or `sed` the contig directly into an output file

> ## Extract contig gff
>```bash
>grep "contig_xxxx" pharokka.gff > contig_xxxx.gff
>
>```
>{: .source}
{: .solution}

> ## Extract contig fasta
>```bash
># if your fasta file is in single-line format
>grep -A 1 "contig_xxxx" phage_contigs.fasta > contig_xxxx.fasta
>
># if you fasta file is in multi-line format
># this tells sed to print between "contig_1026" and the next ">", excluding the last ">"
>sed -n '/contig_1026/,/>/{/>/!p}' phage_contigs.fasta
>
>```
>{: .source}
{: .solution}

Genbank files are a bit more difficult to extract information from but they are formatted sequentially by records. Luckily for us, there are pre-built modules in the biopython library to work with genbank files. [SeqRecord](https://biopython.org/wiki/SeqRecord) and [Bio.SeqIO.parse](https://biopython.org/docs/1.76/api/Bio.SeqIO.html#Bio.SeqIO.parse)

If you `head` the pharokka.gbk file, you might see how the records are structured:  

```bash

LOCUS       contig_1026            35733 bp    DNA     linear   PHG 06-SEP-2024
DEFINITION  contig_1026.
ACCESSION   contig_1026
VERSION     contig_1026
KEYWORDS    .
SOURCE      .
  ORGANISM  .
            .
FEATURES             Location/Qualifiers
     CDS             complement(618..749)
                     /ID="contig_1026_CDS_0001"
                     /transl_table=11
                     /phrog="No_PHROG"
                     /top_hit="No_PHROG"
                     /locus_tag="contig_1026_CDS_0001"
                     /function="unknown function"
                     /product="hypothetical protein"
                     /source="Pyrodigal-gv_0.3.1"
                     /score="3.3"
                     /phase="0"
                     /translation="MTRWNTTLAVYEIWTGTQWQAVASSTYSINYLIVAGGASGAHY*"

See if you can write a script to extract the genbank record for your contig using this module - for clues - see solution below!
```

The whole contig is called a "record" - see the [biopython SeqRecord class](https://biopython.org/wiki/SeqRecord) page for what each element in the record is called, use [Bio.SeqIO.parse](https://biopython.org/docs/1.76/api/Bio.SeqIO.html#Bio.SeqIO.parse) to iterate through these records and [Bio.SeqIO.write](https://biopython.org/docs/1.76/api/Bio.SeqIO.html#Bio.SeqIO.write) to write out the record you want.


> ## extract contig from genbank
>```python
>#!/usr/bin/python3
>
>from Bio import SeqIO
>import sys 
>
># read files in as variables
>input_gbk = sys.argv[1]
>output_gbk = sys.argv[2]
>contig = sys.argv[3]
>
># parse the genbank and write output
>for record in SeqIO.parse(input_gbk, "genbank"):
>    # print(record.id)
>    if record.id == contig:
>        SeqIO.write(record, output_gbk, "genbank")
>```
>{: .source}
{: .solution}

> ## sbatch script to your python script
> ```
>#!/bin/bash
>#SBATCH --tasks=1
>#SBATCH --cpus-per-task=1
>#SBATCH --partition=short,standard,interactive
>#SBATCH --mem=2G
>#SBATCH --time=00:01:00
>#SBATCH --job-name=py_script
>#SBATCH --output=py_script.slurm.%j.out
>#SBATCH --error=py_script.slurm.%j.err
>
># activate the python virtual environment with the packages we need
>source ../py3env/bin/activate
>
># change the output and contig names accordingly
>python3 extract_contig_gbk.py pharokka.gbk output_contig.gbk contig_xxxx
>```
> {: .source}
{: .solution}

> ## Exercise - plot the genome annotations for your pet contig
>
>{:start='4'}
>4. Once we have our fasta, gff and genbank files, we can use [`pharokka_plotter.py`](https://pharokka.readthedocs.io/en/latest/plotting/#pharokka_plotterpy) to plot our pet contig. Feel free to experiment with the different options given in the manual. The base usage is:
>
>```bash
>pharokka_plotter.py -i input.fasta -n pharokka_plot --gff pharokka.gff --genbank pharokka.gbk
>```
> Make sure to add your plot into your lab book!
> {: .source}
{: .challenge}

> ## sbatch script for pharokka_plotter
> ```bash
> #!/bin/bash
>#SBATCH --tasks=1
>#SBATCH --cpus-per-task=10
>#SBATCH --partition=short,standard,interactive
>#SBATCH --mem=10G
>#SBATCH --time=2:00:00
>#SBATCH --job-name=pharokka_1
>#SBATCH --output=pharokka_plotter.slurm.out.%j
>#SBATCH --error=pharokka_plotter.slurm.err.%j
>
>source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh && conda activate pharokka_v1.7.1
>
>#database_path /work/groups/VEO/databases/pharokka/v1.7.1/
>
>echo `date`
>
>pharokka_plotter.py -i contig_1026.fasta -n pharokka_plot_meta_hmm_phanotate --gff 23_meta_hmm_phanotate/contig_1026/contig_1026.gff --genbank 23_meta_hmm_phanotate/contig_1026/contig_1026.gbk
>
>echo `date`
>```
>{: .source}
{: .solution}

The final plot might look something like this:

![pharokka_plot_meta_hmm_phanotate](https://github.com/user-attachments/assets/72e776be-195d-42c5-94d7-8795306482e6)

##  Interpreting your contig

With functional annotation we can begin to get an insight into our viruses to piece together information about their lifestyles, hosts interactions etc, what type of genes they have. You'll find that most genes on viral contigs are not actually annotated - this is because 1) viruses evolve very quickly and assigning gene annotations by homology is difficult, and 2) viruses are highly mosaic - they pick up and lose genes from their hosts and other viruses, meaning not all viruses have all the same genes and viral gene databases do not always account for their variability. The creation of newer tools like [phold](https://github.com/gbouras13/phold?tab=readme-ov-file) aid in this by using structural homologs to assign gene annotation.


> ## Exercise - Interpret your contigs
>
>{:start='5'}
> 5. How many structural genes are annotated?
> 6. Describe at least 2 different types of structural genes
>     a) Where are they placed in your genome?
>     b) To which part of phage structure do they belong? (capsid, tail etc)
>     c) What other viruses have these structural features?
> 7. How many replication genes are annotated?
>     - These could be genes related to transcriptional regulation, DNA/RNA metabolism lysis etc
> 8. Describe the role of the at least 2 replication genes
>     a) What is the function of the gene?
>     b) What part of the replication cycle are they used?
>     c) How do they interact with the host genes (if they do)
> 9. Describe 1 or more interesting genomic features
>     - CRISPR arrays, tRNA, virulence factors etc
> 10. To the best of your ability, infer how this virus might be interacting with its host.
>     - You might investigate clues temperate/lytic lifestyles, auxiliary metabolic genes, anti-crisprs etc
> {: .source}
{: .challenge}
