---
start: True
title: Host Prediction I
teaching: 60
exercises: 60
questions: 
objectives:
    - Understand how biological information is used to predict hosts
    - Understand the difficulties with host prediction
    - Learn about the new techniques that are being used for host prediction
keypoints:
---

{% include base_path.html %}
<p align="center">
    <a href="{{ site.carpentries_site }}"><img src="{{ relative_root_path }}/assets/img/logo_day5.png" alt="Viromics workflow" width="1600" /></a>
</p>

## Host prediction lecture

Listen to lecture by Varada and Malte (pay attention!!!)

> ## Exercise
> 1. What are some biological interactions viruses have with their hosts? (*hint: start with the infection cycle*)
>
> {:start='2'}
> Pick one classical host prediction method:
> 2. In brief, explain how this works
> 3. What is the major challenge with this method?
> 4. How can you be confident in your host prediction?
> 5. Evaluate the following scenarios:
>    a) Two viral contigs match to the same host during the host prediction
>    b) One viral contig matches 2 hosts: [s__Salinibacter ruber](https://gtdb.ecogenomic.org/tree?r=s__Salinibacter%20ruber) and [s__Longimonas halophila](https://gtdb.ecogenomic.org/tree?r=s__Longimonas%20halophila) 
> {: .source}
{: .discussion}

For the hands-on part - we will be using a tool called RaFAH.  [RaFAH paper](https://www.sciencedirect.com/science/article/pii/S2666389921001008?via%3Dihub).  

This tool uses random forest model machine learning models to predict hosts for phages. They train the ML model using manually curated classical host predictions (CRISPR sequences, tRNA and homology based matches). 

