---
start: True
title: "Host Prediction II"
teaching: 120
exercises: 30
questions:
- ""
objectives:
- ""
keypoints:
---

# Host prediction

> ## python script for splitting the assembly into separate files
> ```python
> import os, sys
> from Bio import SeqIO
> 
> def main():
> 
>     assembly_path = os.path.abspath(sys.argv[1])
>     assert assembly_path.endswith(".fasta")
>     
>     out_dir = os.path.abspath(sys.argv[2])
>     if not os.path.exists(out_dir): os.makedirs(out_dir)
> 
>     with open(assembly_path) as handle:
>         for record in SeqIO.parse(handle, "fasta"):
>             out_fasta = os.path.join(out_dir, f"{record.id}.fasta")
>             with open(out_fasta, "w") as fout:
>                 SeqIO.write([record], fout, "fasta")
> 
> if __name__ == "__main__":
>     main()
>```
> {: .source}
{: .solution}

> ## sbatch script for host prediction with RaFAH
> ```bash
> #!/bin/bash
> #SBATCH --tasks=1
> #SBATCH --cpus-per-task=32
> #SBATCH --partition=standard
> #SBATCH --mem=30G
> #SBATCH --time=01:00:00
> #SBATCH --job-name=hostprediction_rafah
> #SBATCH --output=10_results_hostprediction_rafah/hostprediction_rafah.slurm.%j.out
> #SBATCH --error=10_results_hostprediction_rafah/hostprediction_rafah.slurm.%j.err
> 
> # set some variables for running RaFAH on draco
> export PATH=/home/groups/VEO/tools/perl/build/perl-5.32.1/perl-5.32.1:$PATH
> export PERL5LIB=/home/groups/VEO/tools/perl/build/perl-5.32.1/perl-5.32 && 
> rafah='/home/groups/VEO/tools/rafah/RaFAH.pl'
> 
> # RaFAH expects each genome in a separate file. Activate our virtual environment
> # and run a python script to split our filtered assembly into single files
> source ../py3env/bin/activate
> 
> # The script requires the assmbly path and a directory for outputting the contigs
> python 10_split_assembly_file.py ../1.3_virus_identification/20_select_contigs/phage_contigs.fasta 10_results_hostprediction_rafah/split_contigs
> 
> # rafah parameters (https://gensoft.pasteur.fr/docs/RaFAH/0.3/)
> # --predict: run the pipeline for predicting hosts
> # --genomes_dir: the directory with the separate files for each contig
> # --extension: the extension of the contig files 
> perl "$rafah" --predict --genomes_dir 10_results_hostprediction_rafah/split_contigs --extension .fasta
> 
> # technically its not necessary to close the environment, 
> # the session will be terminated after the script finishes.
> deactivate
>```
> {: .source}
{: .solution}
