---
title: Visualizing viral taxonomy
teaching: 0
exercises: 90
questions: []
objectives:
- Plot the network of closely related viruses created by vConTACT3
keypoints:
- The Python package NetworkX can be used to work with networks
---

_This section is suggested as homework._

{% include base_path.html %}
<p align="center">
    <a href="{{ site.carpentries_site }}"><img src="{{ relative_root_path }}/assets/img/contig_164_graph.png" alt="Graph of related phages" width="800" /></a>
</p>

Today, you taxonomically annotated your phage contigs with vConTACT3. 
The tool applies a gene-sharing network approach by computing the number of homologous genes between two viral sequences. 
Based on this information, it integrates unseen contigs into a reference network of taxonomically classified phages, and then uses information about related viruses in the network to infer a taxonomic classification. 
The resulting network, which contains both the reference phages and our assembled contigs, can be found in the output file `graph.cyjs` in your vConTACT3 results folder. 
This file contains the network encoded in JSON format, which can be parsed with the Python standard library package json.

Python also provides several tools for handling and analyzing networks. 
In this section, you will use the NetworkX package to visualize the network of phages. 
The goal is to get an idea of the taxonomic diversity of the contigs in our dataset, and how it compares to reference phages in the database. 
You can either focus on a single contig and its closely related phages (close network neighborhood) or plot the whole network to get an overview of the full virosphere (at least the viruses in the reference database). 

NetworkX provides several [functions for drawing the network](https://networkx.org/documentation/stable/reference/drawing.html). 
Our network only consists of nodes (phages) and weighted edges (number of shared genes) between them, without any layout information (where to position the nodes). 
There are many algorithms for finding a visually pleasing layout for a given network. 
The Kamada-Kawai layout and the spring layout both employ a physical force-based simulation that optimizes the distance or attraction between node pairs. 
In our network, the number of shared genes between two nodes serves as a measure of attraction. 
The spring layout is computationally less demanding and should be used for the visualization of the complete network. 

~~~
# you have to add networkx to the virtual environment first
import networkx as nx
# the json package is part of the standard library
import os, sys, json

# parse the json file and create a networkx graph from it
with open(graph_filename) as json_file:
     graph_json = json.load(json_file)

     # the following lines are necessary for compatibility with networkx
     graph_json["data"] = []
     graph_json["directed"] = False
     for node in graph_json["elements"]["nodes"]:
         # the 'value' attribute will be used for naming the nodes
         node["data"]["value"] = node["data"]["id"]

     # create a networkx graph from the graph_json dictionary
     g = nx.cytoscape_graph(graph_json)

# use draw_networkx_nodes(...) and draw_networkx_edges(...) to draw the graph
# use g.subgraph(...) or g.edge_subgraph(...) to select a subgraph of interest
~~~
{: .language-python}

> ## python script for plotting closely related phages
> ```python
> import os, sys, json
> import networkx as nx
> import matplotlib.pyplot as plt
> 
> def main():
> 
>     graph_filename = os.path.abspath(sys.argv[1])
>     assert graph_filename.endswith(".cyjs")
>     
>     out_filename = os.path.abspath(sys.argv[2])
>     assert out_filename.endswith(".png")
> 
>     contig_id = sys.argv[3]
> 
>     with open(graph_filename) as json_file:
>         graph_json = json.load(json_file)
>         graph_json["data"] = []
>         graph_json["directed"] = False
>         for node in graph_json["elements"]["nodes"]:
>             node["data"]["value"] = node["data"]["id"]
>         g = nx.cytoscape_graph(graph_json)
>         
>     contig_id_node = -1
>     for node in g.nodes(data="name"):
>     	if node[1] == contig_id: contig_id_node = node[0]
>     
>     if contig_id_node == -1: raise(AssertionError(f"Could not find {contig_id} in the graph"))
>     
>     node_set = {contig_id_node}.union({n for n in g.neighbors(contig_id_node)})
>     subgraph =  g.subgraph(node_set)
> 
>     pos = nx.kamada_kawai_layout(subgraph)
>     pos = nx.rescale_layout_dict(pos, 2.0)
>     names = nx.get_node_attributes(subgraph, "name")
>     colors = [('seagreen' if names[n] == contig_id else'lightblue') if names[n].startswith('contig') else 'salmon' for n in subgraph.nodes()]
> 
>     nx.draw_networkx_edges(subgraph, pos=pos, alpha=0.6)
>     nx.draw_networkx_nodes(subgraph, pos=pos, node_size=600, node_color='white')
>     nx.draw_networkx_nodes(subgraph, pos=pos, alpha=0.7, node_size=600, node_color=colors)
>     nx.draw_networkx_labels(subgraph, pos=pos, labels=names, font_size=3, font_weight='bold')
> 
>     ax = plt.gca()
>     ax.axis('off')
>     plt.tight_layout()
>     plt.savefig(out_filename, dpi=800)
> 
> 
> if __name__ == "__main__":
>     main()
>```
> {: .source}
{: .solution}

> ## sbatch script for submitting the plotting script
> ```
> #!/bin/bash
> #SBATCH --tasks=1
> #SBATCH --cpus-per-task=2
> #SBATCH --partition=short
> #SBATCH --mem=5G
> #SBATCH --time=00:10:00
> #SBATCH --job-name=taxonomic_graph
> #SBATCH --output=30_results_taxonomic_graph/taxonomic_graph.slurm.%j.out
> #SBATCH --error=30_results_taxonomic_graph/taxonomic_graph.slurm.%j.err
> 
> # activate the python virtual environment with the packages we need
> source ../py3env/bin/activate
> 
> # in this sbatch script, its not necessarry to create the directory,
> # we already told sbatch to create it for the log files.
> mkdir -p 30_results_taxonomic_graph
> 
> # define an array with contig ids to plot
> declare -a arr=("contig_10" "contig_109", "contig_164")
> 
> # loop through the array
> for contig in "${arr[@]}"
> do
> 	# run our script for plotting the network around a contig of choice
> 	python 30_taxonomic_graph.py ../2.2_viral_taxonomy/02_vcontact3_results_new_contig_names/graph.cyjs 30_results_taxonomic_graph/"$contig"_graph.png $contig
> done
> 
> # deactivate the environment
> deactivate
> ```
> {: .source}
{: .solution}
> 
> ## Plot the network of closely related phages
> You can choose to plot the whole network of all reference phages and contigs, or zoom in and plot a single contig and its neighborhood.
> - Use the taxonomic information encoded in the network to color the nodes.
> - Use draw_networkx_nodes and draw_networkx_edges to better access the plot attributes.
{: .challenge}
