---
title: Viral taxonomy and phylogeny II
teaching: 180
exercises: 0
questions: []
objectives:
- Run vConTACT3 to work out the taxonomy of your viruses
- Interpret this taxonomy
keypoints:
- VConTACT3 and geNomad are two programs used to classify viral taxonomy of sequences using differing strategies
- VConTACT3 uses gene-sharing networks, while GeNomad uses marker genes
- Both methods have strong agreement with the ICTV classifications
---

## What kind of phages exist in your dataset?

For this task, we will use [vConTACT3](https://bitbucket.org/MAVERICLab/vcontact3/src/master/) and [geNomad](https://github.com/apcamargo/genomad). 

### vConTACT3 

vConTACT3 has an underlying assumption that the fraction of shared genes between two viruses represents their evolutionary relationship. 
Thus, the vConTACT3 gene-sharing network closely correlates with the ICTV taxonomy. 

```bash

source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh && conda activate mamba_20231101_python_3.9 

# Need to use with db path? 
vcontact3 --db-path /work/groups/VEO/databases/vcontact3/v20240513/220.json

```

See your options

```bash
vcontact3 run --help
```

Example command
*Note: Take care to have enough memory (RAM) for processing. vConTACT3 loads a full viral network into a graph*

```bash

vcontact3 run --nucleotide virus_genomes.fasta --output output_directory --db-path /work/groups/VEO/databases/vcontact3/v20231101/ -e cytoscape,tree

# other commands you might use
# -t : threads
# --distance-metric : {VirClust,SqRoot,Shorter,Jaccard}
# --min-shared-genes : minimum number of shared genes for two genomes to be connected
# -e : exporting formats/files

```

### Output hints

`final_assignments.csv` contains your taxonomic assignments
`.cyjs` files are network files for Cytoscape
`.pkl.gz` files are compressed pickle files (read with Python)

### geNomad

geNomad has a different way of assigning taxonomy. It uses a marker-gene-to-taxonomy database. 
It first identifies marker genes on the contigs, classifies the taxonomy for each gene, and then consolidates it using weights. 
[Read it here](https://portal.nersc.gov/genomad/taxonomic_assignment.html).

```bash
source /vast/groups/VEO/tools/anaconda3/etc/profile.d/conda.sh

conda activate genNomad_v20230721

```

```bash

genomad annotate [OPTIONS] INPUT OUTPUT DATABASE

```


> ## Exercise
> 1. What percentage of your viruses were classified by each tool?
> 2. How can you judge the accuracy of your taxonomic classifications?
> 3. How many classifications agree/disagree?
> 4. Which viruses are closely related? Which are not?
> 5. What is the maximum taxonomic distance between viruses that are still connected in this graph? 
> 6. What are the limitations of using these tools?
> 
> **Pick one of the viral contigs annotated by vConTACT3**
> 
> 7. How many genes are shared with known reference viruses?
> 8. Which genes are shared with the other known viruses?
> 9. Visualize your virus on the network (see homework exercises [here](https://mgxlab.github.io/Viromics2024/2.2.3_visualizing_taxonomy/index.html)).
> {: .source}
{: .challenge}

> ## Querying pickle files using python
>```python
> import pandas as pd
> 
> # read in the shared genes pickle file
> # this pickle file contains genes shared at even 30% identity - there are other files for 40, 50 ,60 and 70% identity
>df_3 = pd.read_pickle("HMMprofile.0.3_SqRoot_shared_genes.pkl.gz")
>df_3
># check how many genes contig_518 shares with the other contigs
>df_3_contig_518 = df_3.loc[df_3['contig_518'] > 0, 'contig_518']
>df_3_contig_518
> ```
> 
> {: .source}
{: .solution}

## Next steps - getting all your data together

Now that we have results from many different tools and programs, we can build a table to compare and correlate them. You can build this table however you would like - using excel or python etc... If you are using python, I recommend using pandas [`df.merge`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.merge.html) or [`df.join`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html) methods for this task. You will likely need to use it multiple times. Here's a sample of the type of table you need to build:

| Contig   | Contig length (bp) | Coverage | Coverage | Coverage | Completeness | Contamination | Num Genes | Num Genes | Num Genes | Num Genes  | Num Genes  | Num Genes  | Num Genes | Num Genes | Num Genes | Taxonomy                                                        | Taxonomy                                                        | Host Prediction |
| -------- | ------------------ | -------- | -------- | -------- | ------------ | ------------- | --------- | --------- | --------- | ---------- | ---------- | ---------- | --------- | --------- | --------- | --------------------------------------------------------------- | --------------------------------------------------------------- | --------------- |
|          |                    | B62      | B63      | B64      |              |               | CheckV    | CheckV    | CheckV    | Phannotate | Phannotate | Phannotate | GeNomad   | GeNomad   | GeNomad   | vConTACT3                                                       | GeNomad                                                         |                 |
|          |                    |          |          |          |              |               | Total     | Viral     | Bacterial | Total      | Viral      | Bacterial  | Total     | Viral     | Bacterial |                                                                 |                                                                 |                 |
| Contig_1 | 1000               | 5        | 10       | 2        | 97           | 0.3           | 50        | 45        | 2         | 50         | 47         | 2          | 50        | 47        | 2         | Viruses;Duplodnaviria;Heunggongvirae;Uroviricota;Caudoviricetes | Viruses;Duplodnaviria;Heunggongvirae;Uroviricota;Caudoviricetes | unknown         |

 
