---
title: Viral Taxonomy and Phylogeny I
teaching: 60
exercises: 60
questions: 
objectives:
    - Understand the differences between viral and prokaryotic taxonomy
    - Understand the challenges with viral taxonomy and phylogeny
Prerequisites:
- You might want to download and install Cytoscape 
keypoints:
- Viral taxonomy used to be classified by Baltimore Classes but from 2020 onwards, it is based on gennoic content
- Gene sharing networks and marker genes are used for assigned taxonomy
---

Check out the ICTV website: https://ictv.global/

Read the following sections from this seminal paper by [Koonin et al.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7062200/) 
- THE BALTIMORE CLASSES OF VIRUSES, VIRUS HALLMARK GENES, AND MAJOR EVOLUTIONARY TRENDS IN THE VIRUS WORLD

- EVOLUTIONARY LINKS AMONG VIRUSES WITHIN AND ACROSS THE BALTIMORE CLASSES, section: Double-Stranded DNA Viruses

> ## Discussion
>
> ICTV
>     - What is ICTV? (not just a TV channel featuring polar bears)  
>     - What is a realm?  
>     - How many viral genera are there?  
>
> Koonin et al.
>     - What was the Baltimore Classes of Viruses based on?  
>     - What triggered the change in viral taxonomy?  
>     - What is the primary advantage of moving to a genomic based taxonomy?  
>     
> {: .source}
{: .challenge}

How are viruses taxonomically classified?

Given that there is no universal conserved gene, the most typical method involves placement of a viral contig on a gene sharing network. Therefore, more closely related viruses are assumed to share more genes. Genome-to-genome distances are typically calculated based on a summation of the sequence similarity between multiple genes but the formulas on how to do so vary. [VICTOR](https://academic.oup.com/bioinformatics/article/33/21/3396/3933260?login=true) and [vConTACT3](https://bitbucket.org/MAVERICLab/vcontact3/src/master/) work this way.

Newer methods, for example, the use of single copy marker genes within subfamilies and species levels (in the same way as prokaryotic taxonomy is classified) are also being actively developed. [see here for vClassifier](https://www.biorxiv.org/content/10.1101/2024.05.28.596318v1.full) and [geNomad](https://portal.nersc.gov/genomad/pipeline.html)
