---
start: True
title: Viral taxonomy and phylogeny I
teaching: 60
exercises: 60
questions: 
objectives:
    - Understand the differences between viral and prokaryotic taxonomy
    - Understand the challenges with viral taxonomy and phylogeny
Prerequisites:
- You might want to download and install Cytoscape 
keypoints:
- Viral taxonomy used to be classified by Baltimore Classes but from 2020 onwards, it is based on gennoic content
- Gene sharing networks and marker genes are used for assigned taxonomy
---

### How are viruses taxonomically classified?

A widely accepted approach for taxonomic classification is by selecting a _marker gene_ that is shared by all organisms, creating a multiple sequence alignment and phylogenetic tree, and identifying taxa as characteristic branches or lineages in the tree. This approach can be applied to all cellular organisms including bacteria, archaea, and eukaryotes, particularly with marker genes functioning in the ribosome. But only to subgroups of viruses, as no marker gene is universally conserved in all viruses.

There is no single method to classify viruses. Many experts from the global virology community have done their part by classifying viruses according to their specific knowledge. This has generated a patchwork of methods that, ideally, capture the features of different viral lineages and generate meaningful taxa that are in agreement with biology. With the advance of viromics and the discovery of viruses by their genome sequences only, bioinformaticians have developed many methods to cluster them. Thus far, these methods have not been widely adopted for official ICTV virus taxonomy. They are rarely universally applicable, and if they are, there is bound to be some level of disagreement with some taxa that were previously defined by experts.

One popular method in the bacteriophage field is the _gene-sharing network_. This involves creating a network where nodes are genomes and edges represent shared gene families. Tight clusters in this gene-sharing network represent groups of viral genomes that share many genes, and those are interpreted as taxonomic groups. Clusters can be identified with different gene-sharing cutoffs, corresponding to different taxonomic ranks, where closely related viruses (species, genera) share more genes than distant ones (families). Tools like [VICTOR](https://academic.oup.com/bioinformatics/article/33/21/3396/3933260?login=true) and [vConTACT3](https://bitbucket.org/MAVERICLab/vcontact3/src/master/) work this way.

As more and more viruses are sequenced and we get a better view of the _virosphere_, marker genes are making a comeback. Although a universal marker gene shared by all viruses does not exist, marker genes are certainly shared by viruses of lower-ranking taxa (species, genera, families, orders, and in some cases above). This was already clear from the gene-sharing network. Detecting the marker gene is evidence that a virus belongs to a given taxon, and phylogenetic trees can help resolve the lower-level taxonomy, just like ribosomal marker gene trees for cellular organisms. Tools like [vClassifier](https://www.biorxiv.org/content/10.1101/2024.05.28.596318v1.full) and [geNomad](https://portal.nersc.gov/genomad/pipeline.html) use marker gene approaches.

> ## Questions
> Check out the ICTV website: [https://ictv.global/](https://ictv.global/).
> Discuss with your fellow students:
> 1. What is ICTV? (not just a TV channel featuring polar bears)
> 2. How many taxonomic ranks are available to classify the virosphere?
> 3. What is a realm?
> 4. How many viral genera are there?
> 
> Read the following sections from the review ["Global Organization and Proposed Megataxonomy of the Virus World" by Koonin et al.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7062200/):
> - [The Baltimore classes of viruses, virus hallmark genes, and major evolutionary trends in the virus world](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7062200/#s3title)
> - [Evolutionary links among viruses within and across the Baltimore classes, section: Double-Stranded DNA Viruses](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7062200/#s4title:~:text=of%20different%20VHGs.-,Double%2DStranded%20DNA%20Viruses,-The%20dsDNA%20viruses)
> 
> Discuss with your fellow students:
>{:start='5'}
> 5. What are the Baltimore Classes, and what are they based on?
> 6. What triggered the change in viral taxonomy?
> 7. What are the pros and cons of a "genomic taxonomy"?
> {: .source}
{: .challenge}
